### PR TITLE
[Diagnostics] Diagnose missing members via fixes

### PR DESCRIPTION
**Pull request by xedin**
Try to fix constraint system in a way where member
reference is going to be defined in terms of its use,
which makes it seem like parameters match arguments
exactly. Such helps to produce solutions and diagnose
failures related to missing members precisely.

These changes would be further extended to diagnose use
of unavailable members and other structural member failures.

Resolves: rdar://problem/34583132
Resolves: rdar://problem/36989788
Resolved: rdar://problem/39586166
Resolves: rdar://problem/40537782
Resolves: rdar://problem/46211109

### CODE REVIEW

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -3781,16 +3798,59 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
...
+
+    // FIXME(diagnostics): This is more of a hack than anything.
+    // Let's not try to suggest that there is no member related to an
+    // obscure underscored type, the real problem would be somewhere
+    // else. This helps to diagnose pattern matching cases.
+    {
+      if (auto *metatype = baseTy->getAs<MetatypeType>()) {
+        auto instanceTy = metatype->getInstanceType();
+        if (auto *NTD = instanceTy->getAnyNominal()) {
+          if (NTD->getName() == getASTContext().Id_OptionalNilComparisonType)
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: You can use .is(...) instead of == getIdentifier(...)

File: lib/Sema/CSDiagnostics.cpp [hunk 2]
```diff
@@ -1332,3 +1333,146 @@ bool SubscriptMisuseFailure::diagnoseAsNote() {
...
+    }
+  } else if (auto moduleTy = baseType->getAs<ModuleType>()) {
+    emitDiagnostic(baseExpr->getLoc(), diag::no_member_of_module,
+                   moduleTy->getModule()->getName(), Name)
+        .highlight(baseExpr->getSourceRange())
+        .highlight(nameLoc.getSourceRange());
+    return true;
+  } else {
+    // Check for a few common cases that can cause missing members.
+    if (baseType->is<EnumType>() && Name.isSimpleName("rawValue")) {
```
Comment (slavapestov) [inline lib/Sema/CSDiagnostics.cpp]: Maybe getEnumOrBoundGenericEnum instead?

File: lib/Sema/CSDiagnostics.cpp [hunk 3]
```diff
@@ -1332,3 +1333,146 @@ bool SubscriptMisuseFailure::diagnoseAsNote() {
...
+
+/// When a user refers a enum case with a wrong member name, we try to find a
+/// enum element whose name differs from the wrong name only in convention;
+/// meaning their lower case counterparts are identical.
+///   - DeclName is valid when such a correct case is found; invalid otherwise.
+DeclName MissingMemberFailure::findCorrectEnumCaseName(
+    Type Ty, TypoCorrectionResults &corrections, DeclName memberName) {
+  if (memberName.isSpecial() || !memberName.isSimpleName())
+    return DeclName();
+  if (!Ty->is<EnumType>() && !Ty->is<BoundGenericEnumType>())
```
Comment (slavapestov) [inline lib/Sema/CSDiagnostics.cpp]: You can use getEnumOrBoundGenericEnum()

**Review by slavapestov — APPROVED** [review]
Awesome!

File: lib/Sema/CSDiagnostics.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Thanks! This code come partially from CSDiag so I didn't revisit it completely...

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Fixed!

File: lib/Sema/CSDiagnostics.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Fixed!

File: lib/Sema/CSDiagnostics.cpp (hunk 3 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Fixed!

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Right, that’s what I thought too, I also looked at other uses of Id_* cached identifiers

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Pull request merged**